common-components/proto/api/type/decimal.proto

```diff
@@ -0,0 +1,36 @@
+syntax = "proto3";
+
+package api.type;
+
+option cc_enable_arenas = true;
+option go_package = "aip.dev/type/decimal;decimal";
+option java_multiple_files = true;
+option java_outer_classname = "DecimalProto";
+option java_package = "dev.aip.type";
+option objc_class_prefix = "AIPTYPE";
+
+// Represents a decimal number in a form similar to scientific notation.
+//
+// Examples:
+// 17            === {digits: 17   exponent: 0} (or just {digits: 17})
+// -0.005        === {digits: -5   exponent: -3}
+// 33.5 million  === {digits: 335  exponent: 5}
+// 11/8 (1.375)  === {digits: 1375 exponent: -3}
+message Decimal {
+  // The significant digits (a.k.a. significand or mantissa) of the number.
+  int64 digits = 1;
```

Comment:
Either base or mantissa would be preferred over "digits".

Honestly I think mantissa is the ideal choice. Yes, nobody knows the word, but it's precise and checking Merriam-Webster or Google is not hard.

Reply:
Yes, people can always look up a definition, but that certainly has worse ergonomics than something more intuitive.


Wikipedia says (emphasis mine):

> The significand[1] (also mantissa[2] or coefficient,[1] sometimes also argument, or ambiguously fraction[3][4][nb 1] or characteristic[5][2]) is part of a number in scientific notation or in floating-point representation, **consisting of its significant digits**.

Since people Googling "mantissa" will likely end up reading that paragraph and going "oh, it means the significant digits", just naming this field "digits" (or "significant_digits") shortcuts that process with the same end result.

I am happy to be outvoted on this, but I'd like to solicit more opinions.

Reply:
`significant_digits` is actually an interesting idea.

I think the real way to decide is to see what common programming language implementations use. I don't actually know, so let's find out really quick (note: I'll use snake case regardless of language conventions)...

- C++: No library found.
- C#: `lo` / `mid` / `high` (mostly non-public)
- Go: `int_part`
- Java: `unscaled_value`
- PHP: `value`
- Python: `digits`
- Rust: `mantissa`
- Ruby: I can't seem to find a public API term for this in Ruby BigDecimal. Unfortunately, they _do_ use digits, but not as the actual digits of the number, but rather as the _number of significant digits_ (so, e.g. for 3.14159, Ruby `digits` would be 6).

Ooof. _This_ hurts.

Reply:
I think I'm persuaded that `digits` is a bad idea (the Ruby use scares me). But I could be easily persuaded to something else like `unscaled_value`.

I also now oppose `significant_digits` for the Ruby reason, too.

Reply:
Maybe `significand`?

Reply:
Yeah, I found something similar - there's no consistent precedent set by major languages, which is why I decided to propose something from first principles.

I don't think `significand` is any better or worse than `mantissa`; I'm currently more or less neutral between the two.

I think you have a totally fair and reasonable perspective.  I am going to solicit other opinions in the weekly meeting tomorrow though.

Reply:
Thanks for the feedback; changed.